lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,331 @@
+	payload, err := utils.GetAndReplaceRequestBody(req)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	reqCopy := req.Clone(ctx)
+	reqCopy.RequestURI = ""
+	reqCopy.Body = io.NopCloser(bytes.NewReader(payload))
+
+	// Connection is hop-by-hop header, drop.
+	reqCopy.Header.Del("Connection")
+
+	// force HTTPS, set host URL.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = e.sessionCtx.Database.GetURI()
+	reqCopy.Host = e.sessionCtx.Database.GetURI()
+
+	return reqCopy, nil
+}
+
+// emitAuditEvent writes the request and response to audit stream.
+func (e *Engine) emitAuditEvent(req *http.Request) {
+	// Try to read the body and JSON unmarshal it.
+	// If this fails, we still want to emit the rest of the event info; the request event Body is nullable,
+	// so it's ok if body is nil here.
+	body, err := libaws.UnmarshalRequestBody(req)
+	if err != nil {
+		e.Log.WithError(err).Warn("Failed to read request body as JSON, omitting the body from the audit event.")
+	}
+
+	ev := &apievents.OpenSearchRequest{
+		Metadata: common.MakeEventMetadata(e.sessionCtx,
+			events.DatabaseSessionOpenSearchRequestEvent,
+			events.OpenSearchRequestCode),
+		UserMetadata:     common.MakeUserMetadata(e.sessionCtx),
+		SessionMetadata:  common.MakeSessionMetadata(e.sessionCtx),
+		DatabaseMetadata: common.MakeDatabaseMetadata(e.sessionCtx),
+		Method:           req.Method,
+		Path:             req.URL.Path,
+		RawQuery:         req.URL.RawQuery,
+		Body:             body,
+	}
+
+	e.Audit.EmitEvent(req.Context(), ev)
+}
+
+// sendResponse sends the response back to the OpenSearch client.
+func (e *Engine) sendResponse(serverResponse *http.Response) error {
+	payload, err := utils.GetAndReplaceResponseBody(serverResponse)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// serverResponse may be HTTP2 response, but we should reply with HTTP 1.1
+	clientResponse := &http.Response{
+		ProtoMajor:    1,
+		ProtoMinor:    1,
+		StatusCode:    serverResponse.StatusCode,
+		Body:          io.NopCloser(bytes.NewBuffer(payload)),
+		Header:        serverResponse.Header.Clone(),
+		ContentLength: int64(len(payload)),
+	}
+
+	if err := clientResponse.Write(e.clientConn); err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// checkAccess does authorization check for OpenSearch connection about
+// to be established.
+func (e *Engine) checkAccess(ctx context.Context) error {
+	authPref, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	state := e.sessionCtx.GetAccessState(authPref)
+	dbRoleMatchers := role.DatabaseRoleMatchers(
+		e.sessionCtx.Database,
+		e.sessionCtx.DatabaseUser,
+		e.sessionCtx.DatabaseName,
+	)
+	err = e.sessionCtx.Checker.CheckAccess(
+		e.sessionCtx.Database,
+		state,
+		dbRoleMatchers...,
+	)
+
+	if e.sessionCtx.Identity.RouteToDatabase.Username == "" {
+		return trace.BadParameter("database username required for OpenSearch")
+	}
+
```

Comment:
curious why this check is needed or if it should be checked earlier?

Reply:
I was just surprised that an empty database user can pass the role matchers check above. Is the intent here to return a bad parameter instead of access denied? Since `err` is returned a few lines above but the `err` is not checked after this if.